src/comgr.cpp

```diff
@@ -0,0 +1,490 @@
+                    AMD_COMGR_ACTION_ASSEMBLE_SOURCE_TO_RELOCATABLE,
+                    AMD_COMGR_ACTION_DISASSEMBLE_RELOCATABLE_TO_SOURCE,
+                    AMD_COMGR_ACTION_DISASSEMBLE_EXECUTABLE_TO_SOURCE,
+                    AMD_COMGR_ACTION_DISASSEMBLE_BYTES_TO_SOURCE,
+                    AMD_COMGR_ACTION_COMPILE_SOURCE_TO_FATBIN);
+    return oss.str();
+}
+
+static bool PrintVersion()
+{
+    std::size_t major = 0;
+    std::size_t minor = 0;
+    (void)amd_comgr_get_version(&major, &minor);
+    MIOPEN_LOG_NQI("comgr v." << major << '.' << minor);
+    return true;
+}
+
+static std::string GetStatusText(const amd_comgr_status_t status)
+{
+    const char* reason = nullptr;
+    if(AMD_COMGR_STATUS_SUCCESS != amd_comgr_status_string(status, &reason))
+        reason = "<Unknown>";
+    return std::string(reason) + " (" + std::to_string(static_cast<int>(status)) + ')';
+}
+
+static void LogOptions(const char* options[], size_t count)
+{
+    if(miopen::IsEnabled(MIOPEN_DEBUG_COMGR_LOG_OPTIONS{}) &&
+       miopen::IsLogging(miopen::LoggingLevel::Info))
+    {
+        std::ostringstream oss;
+        for(std::size_t i = 0; i < count; ++i)
+            oss << options[i] << '\t';
+        MIOPEN_LOG_I(oss.str());
+    }
+}
+
+class Dataset;
+static std::string GetLog(const Dataset& dataset, bool comgr_error_handling = false);
+
+struct ComgrError : std::exception
+{
+    amd_comgr_status_t status;
+    std::string text;
+
+    ComgrError(const amd_comgr_status_t s) : status(s) {}
+    ComgrError(const amd_comgr_status_t s, const std::string& t) : status(s), text(t) {}
+    const char* what() const noexcept override { return text.c_str(); }
+};
+
+[[noreturn]] static void Throw(const amd_comgr_status_t s) { throw ComgrError{s}; }
+[[noreturn]] static void Throw(const amd_comgr_status_t s, const std::string& text)
+{
+    throw ComgrError{s, text};
+}
+
+struct ComgrOwner
+{
+    ComgrOwner(const ComgrOwner&) = delete;
+
+    protected:
+    ComgrOwner() {}
+    ComgrOwner(ComgrOwner&&) = default;
+};
+
+class Data : ComgrOwner
+{
+    amd_comgr_data_t handle = {0};
+    friend class Dataset; // for GetData
+    Data(amd_comgr_data_t h) : handle(h) {}
+
+    public:
+    Data(amd_comgr_data_kind_t kind) { ECI_THROW(amd_comgr_create_data(kind, &handle), kind); }
+    Data(Data&&) = default;
+    ~Data() { EC(amd_comgr_release_data(handle)); }
+    auto operator()() const { return handle; }
+    void SetName(const std::string& s) const
+    {
+        ECI_THROW(amd_comgr_set_data_name(handle, s.c_str()), s);
+    }
+    void SetBytes(const std::string& bytes) const
+    {
+        ECI_THROW(amd_comgr_set_data(handle, bytes.size(), bytes.data()), bytes.size());
+    }
+
+    private:
+    std::size_t GetSize() const
+    {
+        std::size_t sz;
+        EC_THROW(amd_comgr_get_data(handle, &sz, nullptr));
+        return sz;
+    }
+
+    public:
+    std::size_t GetBytes(std::vector<char>& bytes) const
```

Comment:
This should return `std::vector<char>` and not use output parameters.

Reply:
Unfortunately NRVO is inevitable in this case. This means that compiler is not required to skip copy and move, so the code might be less efficient. Worse is that Debug and Release versions would reveal different behavior (incl. in the time domain) because NRVO is an optimization.

Current code is an old style but bullet-proof; copy and move would never happen.